source/contributing/implementation/datamanipulator.rst

```diff
@@ -0,0 +1,467 @@
+``org.spongepowered.common.data.key.KeyRegistry`` class and find the static ``generateKeyMap()`` function.
+There add a line to register (and create) your used keys.
+
+.. code-block:: java
+
+    public static void registerKeys() {
+        keyMap.put("max_health", makeSingleKey(Double.class, MutableBoundedValue.class, of("MaxHealth")));
+    }
+
+The ``keyMap`` maps strings to ``Key``\ s. The string used should be the corresponding constant name from
+the ``Keys`` utility class in lowercase. The ``Key`` itself is created by one of the static methods
+provided by ``KeyFactory``, in most cases ``makeSingleKey``. ``makeSingleKey`` requires first a class reference
+for the underlying data, which in our case is a "Double", then a class reference for the ``Value`` type used.
+The third argument is the ``DataQuery`` used for serialization. It is created from the statically imported
+``DataQuery.of()`` method accepting a string. This string should also be the constant name, stripped of
+underscores and capitalization changed to upper camel case.
+
+
+4. Implement the DataProcessors
+===============================
+
+Next up is the ``DataProcessor``. A ``DataProcessor`` serves as a bridge between our ``DataManipulator`` and
+Minecraft's objects. Whenever any data is requested from or offered to ``DataHolders`` that exist in Vanilla
+Minecraft, those calls end up being delegated to a ``DataProcessor`` or a ``ValueProcessor``.
+
+For your name, you should use the name of the ``DataManipulator`` interface and append ``Processor``. Thus for ``HealthData`` we create a ``HealthDataProcessor``.
+
+In order to reduce boilerplate code, the ``DataProcessor`` should inherit from the appropriate abstract class in
+the ``org.spongepowered.common.data.processor.common`` package. Since health can only be present on certain
+entities, we can make use of the ``AbstractEntityDataProcessor`` which is specifically targeted at ``Entities``
+based on ``net.minecraft.entity.Entity``. ``AbstractEntitySingleDataProcessor`` would require less
+implementation work, but cannot be used as ``HealthData`` contains more than just one value.
+
+.. code-block:: java
+
+    public class HealthDataProcessor extends AbstractEntityDataProcessor<EntityLivingBase, HealthData, ImmutableHealthData> {
+        public HealthDataProcessor() {
+            super(EntityLivingBase.class);
+        }
+        [...]
+    }
+
+Depending on which abstraction you use, the methods you have to implement may differ greatly, depending on how
+much implementation work already could be done in the abstract class. Generally, the methods can be categorized.
+
+.. tip::
+
+    It is possible to create multiple ``DataProcessor``\ s for the same data. If vastly different ``DataHolder``\ s
+    should be supported (for example both a ``TileEntity`` and a matching ``ItemStack``), it may be beneficial to
+    create one processor for each type of ``DataHolder`` in order to make full use of the provided abstractions.
```

Comment:
placement of these specialized processors should follow the package structure of `item`/`tileentity`/`entity`.